daffodil-runtime1/src/main/scala/edu/illinois/ncsa/daffodil/processors/parsers/IBM4690PackedDecimalParsers.scala

```diff
@@ -0,0 +1,108 @@
+import edu.illinois.ncsa.daffodil.schema.annotation.props.gen.LengthUnits
+import edu.illinois.ncsa.daffodil.schema.annotation.props.gen.YesNo
+import edu.illinois.ncsa.daffodil.processors.ElementRuntimeData
+import edu.illinois.ncsa.daffodil.util.DecimalUtils._
+import java.lang.{ Long => JLong, Number => JNumber }
+import edu.illinois.ncsa.daffodil.processors.ParseOrUnparseState
+
+class IBM4690PackedDecimalKnownLengthParser(e: ElementRuntimeData, signed: YesNo, binaryDecimalVirtualPoint: Int, val lengthInBits: Int)
+  extends IBM4690PackedDecimalParserBase(e, signed, binaryDecimalVirtualPoint)
+  with HasKnownLengthInBits {
+}
+
+class IBM4690PackedDecimalRuntimeLengthParser(val e: ElementRuntimeData, signed: YesNo, binaryDecimalVirtualPoint: Int, val lengthEv: Evaluatable[JLong], val lUnits: LengthUnits)
+  extends IBM4690PackedDecimalParserBase(e, signed, binaryDecimalVirtualPoint)
+  with HasRuntimeExplicitLength {
+}
+
+abstract class IBM4690PackedDecimalParserBase(e: ElementRuntimeData, signed: YesNo, binaryDecimalVirtualPoint: Int)
+  extends PrimParserObject(e) {
+
+  protected def getBitLength(s: ParseOrUnparseState): Int
+
+  def parse(start: PState): Unit = {
+    val nBits = getBitLength(start)
+    val dis = start.dataInputStream
+
+    if (!dis.isDefinedForLength(nBits)) {
+      PE(start, "Insufficient bits in data. Needed %d bit(s) but found only %d available.", nBits, dis.remainingBits.get)
+      return
+    }
+
+    val bigDec = IBM4690ToBigDecimal(dis.getByteArray(nBits), binaryDecimalVirtualPoint)
+    start.simpleElement.overwriteDataValue(bigDec)
+  }
+}
+
+class IBM4690PackedIntegerRuntimeLengthParser(val e: ElementRuntimeData, signed: Boolean, val lengthEv: Evaluatable[JLong], val lUnits: LengthUnits)
+  extends IBM4690PackedIntegerBaseParser(e, signed)
+  with HasRuntimeExplicitLength {
+}
+
+class IBM4690PackedIntegerKnownLengthParser(e: ElementRuntimeData, signed: Boolean, val lengthInBits: Int)
+  extends IBM4690PackedIntegerBaseParser(e, signed)
+  with HasKnownLengthInBits {
+}
+
+abstract class IBM4690PackedIntegerBaseParser(e: ElementRuntimeData, signed: Boolean)
+  extends PrimParserObject(e) {
+
+  protected def getBitLength(s: ParseOrUnparseState): Int
+
+  def parse(start: PState): Unit = {
+    val nBits = getBitLength(start)
+    if (nBits == 0) return // zero length is used for outputValueCalc often.
+    val dis = start.dataInputStream
+
+    if (!dis.isDefinedForLength(nBits)) {
+      PE(start, "Insufficient bits in data. Needed %d bit(s) but found only %d available.", nBits, dis.remainingBits.get)
+      return
+    }
+
+    val int: JNumber =
+      if (signed) {
+        if (nBits > 64) { IBM4690ToBigInteger(dis.getByteArray(nBits)) }
+        else { IBM4690ToBigInteger(dis.getByteArray(nBits)).longValue() }
+      } else {
+        if (nBits >= 64) { IBM4690ToBigInteger(dis.getByteArray(nBits)) }
+        else { IBM4690ToBigInteger(dis.getByteArray(nBits)).longValue() }
+      }
```

Comment:
I don't think if (signed) is being used here correctly. If if/else branches are doing basically the same thing. How does signedness apply to these zoned/packed types?